src/dapp/extensionProvider.ts

```diff
@@ -0,0 +1,214 @@
+  }
+
+  private openExtensionPopup() {
+    if (!this.initialized) {
+      throw new Error("Wallet provider is not initialised, call init() first");
+    }
+    this.extensionPopupWindow = window.open(
+      this.extensionURL,
+      this.popupName,
+      this.popupOptions
+    );
+  }
+
+  private startBgrMsgChannel(
+    operation: string,
+    connectData: any
+  ): Promise<any> {
+    return new Promise((resolve, reject) => {
+      window.postMessage(
+        {
+          target: "erdw-inpage",
+          type: operation,
+          data: connectData,
+        },
+        window.origin
+      );
+
+      const eventHandler = (event: any) => {
+        if (
+          event.isTrusted &&
+          event.data.type &&
+          event.data.target === "erdw-contentScript"
+        ) {
+          switch (event.data.type) {
+            case "logoutResponse":
+              window.removeEventListener("message", eventHandler);
+              resolve(true);
+              break;
+          }
+        }
+      };
+      setTimeout(() => {
+        reject(
+          "Extension logout response timeout. No response from extension."
+        );
+      }, 3000);
+      window.addEventListener("message", eventHandler, false);
+    });
+  }
+
+  private startExtMsgChannel(operation: string, connectData: any): any {
+    return new Promise((resolve, reject) => {
+      var isResolved = false;
+      const eventHandler = (event: any) => {
+        if (
+          event.isTrusted &&
+          event.data.type &&
+          event.data.target === "erdw-extension"
+        ) {
+          switch (event.data.type) {
+            case "popupReady":
+              event.ports[0].postMessage({
+                target: "erdw-inpage",
+                type: operation,
+                data: connectData,
+              });
+              break;
+            case "connectResult":
+              this.extensionPopupWindow?.close();
+              this.account = event.data.data;
+              window.removeEventListener("message", eventHandler);
+              isResolved = true;
+              resolve(event.data.data);
+              break;
+
+            default:
+              this.handleExtResponseErr(event);
+              this.extensionPopupWindow?.close();
+              window.removeEventListener("message", eventHandler);
+              isResolved = true;
+              resolve(event.data.data);
+              break;
+          }
+        }
+      };
+      var windowCloseInterval = setInterval(() => {
+        if (this.extensionPopupWindow?.closed) {
+          window.removeEventListener("message", eventHandler);
+          clearInterval(windowCloseInterval);
+          if (!isResolved)
+            reject("Extension window was closed without response.");
+        }
+      }, 500);
+
+      window.addEventListener("message", eventHandler, false);
+    });
+  }
+
+  private handleExtResponseErr(event: any) {
+    if (!event.data || !event.data.data) {
```

Comment:
Is this condition correct?
`if (!event.data) {...we-are-checking event.data.type}`

I also think it will be cleaner with early exits (less conditions and indentation)
```js
if (!havingCorrectCondition) {
    return;
}

if (....) {
    throw;
}
```